File: apps/android/core/common/src/main/kotlin/app/gauja/core/common/Secret.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.common

// Only explicit transport/storage boundaries may borrow the bytes. Never a saved-state value.
class Secret(bytes: ByteArray) {
    private val value = bytes.copyOf()

    fun <T> useBytes(block: (ByteArray) -> T): T {
        val copy = value.copyOf()
        return try {
            block(copy)
        } finally {
            copy.fill(0)
        }
    }

    override fun toString(): String = "[REDACTED]"
}

File: apps/ios/Packages/Common/Sources/Common/Secret.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import Foundation

// Intentionally not Codable; wire encoding and secure persistence are explicit boundaries.
public struct Secret: Sendable, CustomStringConvertible, CustomDebugStringConvertible, CustomReflectable {
    private let bytes: Data
    public init(_ bytes: Data) { self.bytes = bytes }

    public func withBytes<Value>(_ body: (Data) throws -> Value) rethrows -> Value { try body(bytes) }
    public var description: String { "[REDACTED]" }
    public var debugDescription: String { description }
    public var customMirror: Mirror { Mirror(self, children: [:]) }
}

extension Secret {
    public func utf8() throws -> String {
        try withBytes {
            guard let value = String(data: $0, encoding: .utf8) else { throw AppError.validation }
            return value
        }
    }
}
